src/meltano/core/upgrade_service.py

```diff
@@ -23,6 +25,48 @@
     from meltano.core.project import Project
 
 
+def _get_pep610_data() -> dict[str, t.Any] | None:
+    dist = distribution("meltano")
+    if contents := dist.read_text("direct_url.json"):
+        return json.loads(contents)
+
+    return None
+
+
+def _check_editable_installation(*, force: bool) -> None:
+    if pep610_data := _get_pep610_data():
+        url: str | None = pep610_data.get("url")
+        dir_info: dict[str, t.Any] = pep610_data.get("dir_info", {})
+        if (  # pragma: no branch
+            url and dir_info and dir_info.get("editable", False) and not force
+        ):
+            meltano_dir = url.removeprefix("file://")
+            raise AutomaticPackageUpgradeError(
+                reason="it is installed from source",
+                instructions=f"navigate to `{meltano_dir}` and run `git pull`",
+            )
```

Comment:
I find it easier to follow when we exit early in response to cases such as `_get_pep610_data()` being `None`

```suggestion
    if (pep610_data := _get_pep610_data()) is None:
        return None
    url: str | None = pep610_data.get("url")
    dir_info: dict[str, t.Any] = pep610_data.get("dir_info", {})
    if (  # pragma: no branch
        url and dir_info and dir_info.get("editable", False) and not force
    ):
        meltano_dir = url.removeprefix("file://")
        raise AutomaticPackageUpgradeError(
            reason="it is installed from source",
            instructions=f"navigate to `{meltano_dir}` and run `git pull`",
        )
```